### PR TITLE
schemachanger: use bazel test sharding for comparator tests

### LINKED ISSUE #111978: pkg/sql/schemachanger/schemachanger_test: TestComparatorFromLogicTests failed
pkg/sql/schemachanger/schemachanger_test.TestComparatorFromLogicTests [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SchemaChangerComparatorFromSqlLogicStatementsBazel/12090674?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SchemaChangerComparatorFromSqlLogicStatementsBazel/12090674?buildTab=artifacts#/) on master @ [981b050f28f406b495b530ec5725fa2c91cd20c5](https://github.com/cockroachdb/cockroach/commits/981b050f28f406b495b530ec5725fa2c91cd20c5):

```
    test_server_shim.go:124: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    comparator.go:98: Executing "CREATE DATABASE IF NOT EXISTS test"
    comparator.go:98: Executing "CREATE USER testuser"
    comparator.go:98: Executing "CREATE TABLE regression_81609 (g GEOMETRY NOT NULL, INVERTED INDEX (g))"
    comparator.go:98: Executing "INSERT INTO regression_81609 SELECT st_rotatex('0105000000060000000102000000020000002CF565DB1790F041C2509EC57231F2411C1E119D614AE7C166AFF2D39B030142010200000003000000809745B2972CF84160399714F897E241E0D03302F1C6D6410EFC89BE9FE20042F83C7FD5C4ACD6411854CBFDFF42DE41010200000004000000C4A695BE1FC2FD411CE9BCE66C42D8C142F59F55FD7800424C0D1238165FE041A07574B08DC0D94170D034F5A969EB41D843A641FF49E741B6F0864FF80E0242010200000004000000781D60294113F2412C338441624C02C2100451CE196BED41525C0281389BEEC11430818C69A2EF415C09A4FD752D024200E4B02D00CB9341C047BBFE74DDE44101020000000400000040B0B6B74C96A0C162D23CFAACB8E7C11CD943F294E9E5C187C18248646102C2709F7E29F32AC4C160D1A56EC799FAC1607737DC4662E941ADC9125B944FF6C1010200000004000000F097AF975517F441714AF1717A16F0C15442BA391710F841BEBE416B6780F241E01C9C5F7602B6413CC620A0C9C3E14140E714D68300E2C1A0C4038270B4BAC1':::GEOMETRY::GEOMETRY, '-Inf':::FLOAT8)::GEOMETRY"
    comparator.go:98: Executing "SELECT st_asencodedpolyline(geomfromewkb(('01070000A0E61000000300000001060000A0E6100000040000000103000080010000000B0000006F9AF3F7D81863C057397A0B3ACD42C006ED79D13D2DF24130E1CBF62A4A45C01B6DC0D279F452C022F8C672D2C1F9C140DB9D23050141C0B53FBA6E396656C0C0A1842E2B53EC413A8DC94C568D61401BEF0352091651C05C681DE175D2FD41A048C859DF7C5440C00E327AC97FF63F68BBA8144D0EF8416E6042F340DD6040805791CE56602740DC6C3922676401C29C6D4009B1334E40625E855D7A7E5140D4241F7D86D2D5C14B1E4BCA077A58C07806BF9D4D084D4074E74BE4DCAA00C2C11D020045A051C0D81F8E3DD0AE3740B8315FF7F63AF4C1BAD97494BA9860C07411EFAF0A6C3D4038E26DE7A194F4C16F9AF3F7D81863C057397A0B3ACD42C006ED79D13D2DF2410103000080010000000B000000610D90AE22A55FC0B8265672F7562940CE3E91EB7A320042CC96C05617B649C085229B1A09E153C0D84FA7998764EB4190FEB5A7C71B2DC03EBBCB6B588F4DC02891E1ED4A7DE6417C0F7BBC14C75840E08D4D210E2F21C04180E453EDEDFFC112E57185B2435240F0D20FC2D7C11640F8BF2E3AB98DD2C1D0A9A94D6EA14140FCABD1E5BF074B400010C2F75AA16F4166BC301BE6FB55C002F9F182DE3D53400038B551BC0495C199463681FA6E5CC0184D3E0AE179554030323F0F2525F4415791FBE04FEA61C0361E8733C77A5240C84AEBEAD466EEC1D5012A75E04F63C0A8BF403F623D4240182CFEFA9071E0C1610D90AE22A55FC0B8265672F7562940CE3E91EB7A3200420103000080010000000C000000004CADA6B302E0BF3EE5FA281ACE3BC0E051ABD6580901429B65A22CCBDC53C003DEBC27260E4AC07E45691D3E7AF541408E6FAB9A1008C016E33E2631FD40C0EEF80F0FF3E8F341A8C5EAA8B0C836409E85B8193CE952C0F4448FFF9B31F24190F9A9FCD2AC604004F99C51ED4356C0903A1F652902D4C1C01B737BD389554080315427550738C05C51AE075475F64124F6F92382235C40148AB5FED33D38C048960D3AC6D7E8412C2E310EABA5514058EA10C427562EC0A82D4D1A69D7FE41621FC77070C26540FE84B78FE5BE51400962385E53FAFEC1483899F770B14DC020BD4BBBAB604D401C08306EA143E84140337962764325C0002F089A1DF3F83FAC9B66F32C2BE841004CADA6B302E0BF3EE5FA281ACE3BC0E051ABD65809014201030000800100000006000000D03FF366A2C95C40E59BFCECDD7553C0D028B038BB63D041A40AEF5CB0195840006FB928107F084020CA0A39A4AED34100F09D4126336140E8AC65DAC4954C4024406D4040D1EE41F04B3F3DA6C75740106594C0ADCE4440A498B6372039FD4155D08A1E777452C084CC004170E949409C87DF53EB9AFDC1D03FF366A2C95C40E59BFCECDD7553C0D028B038BB63D04101020000A0E61000000200000074F55D784E5F50408F7DE14351CD41C0FCCCE7B1B166F541583C392FCD7554409022EAF3EFCC39C03CDE1C0E29A7EB4101040000A0E6100000060000000101000080E8801DFC8A2130C04401B4AE182F50C0BC13340EB7E2E1C1010100008010DC3F808D824D406C9927CDAE5144409A464D93BBD7F94101010000806F0682A50FE253C030E1964D47F74BC00E752ECB60B0F441010100008014F5242FC2C444C01257DB8E51FA51404449E2FB7D2AFFC1010100008069945BA1FD2E50C0501CA74B7A661C40CB83CB9A4894F1C10101000080C0BBBC345A4B51C0E08F6EBB2F250E400681A78F4FFCF741'::GEOGRAPHY)::BYTES)::GEOMETRY)"
    schemachanger_test.go:922: schema changer comparator testing succeeded on logic test "geospatial_regression"
=== RUN   TestComparatorFromLogicTests/lookup_join_spans
    schemachanger_test.go:917: running schema changer comparator testing on statements collected from logic test "lookup_join_spans"
    comparator.go:98: Executing "CREATE DATABASE IF NOT EXISTS test"
    comparator.go:98: Executing "CREATE USER testuser"
    comparator.go:98: Executing "CREATE TABLE metrics (id SERIAL8 PRIMARY KEY, nullable INT8, name STRING, INDEX name_index (name))"
    comparator.go:98: Executing "INSERT INTO metrics(id, nullable, name) VALUES (1, NULL, 'cpu'), (2, 1, 'cpu'), (3, NULL, 'mem'), (4, 2, 'disk')"
    comparator.go:98: Executing "CREATE TABLE metric_values (metric_id INT8, \"time\" TIMESTAMPTZ, nullable INT8, value INT8, PRIMARY KEY (metric_id, \"time\"), INDEX secondary (metric_id, nullable, \"time\"))"
    comparator.go:98: Executing "INSERT INTO metric_values(metric_id, \"time\", nullable, value) VALUES (1, '2020-01-01 00:00:00+00:00', NULL, 0), (1, '2020-01-01 00:00:01+00:00', 1, 1), (2, '2020-01-01 00:00:00+00:00', NULL, 2), (2, '2020-01-01 00:00:01+00:00', 2, 3), (2, '2020-01-01 00:01:01+00:00', -11, 4), (2, '2020-01-01 00:01:02+00:00', -10, 5), (3, '2020-01-01 00:01:00+00:00', NULL, 6), (3, '2020-01-01 00:01:01+00:00', 3, 7)"
    comparator.go:98: Executing "CREATE TABLE metric_values_desc (metric_id INT8, \"time\" TIMESTAMPTZ, nullable INT8, value INT8, PRIMARY KEY (metric_id, \"time\" DESC), INDEX secondary (metric_id, nullable DESC, \"time\" DESC))"
    comparator.go:98: Executing "INSERT INTO metric_values_desc SELECT * FROM metric_values"
    comparator.go:98: Executing "ALTER TABLE metric_values INJECT STATISTICS e'[\\n{\\n\"columns\": [\"metric_id\"],\\n\"created_at\": \"2018-01-01 1:00:00.00000+00:00\",\\n\"row_count\": 1000,\\n\"distinct_count\": 10\\n},\\n{\\n\"columns\": [\"time\"],\\n\"created_at\": \"2018-01-01 1:30:00.00000+00:00\",\\n\"row_count\": 1000,\\n\"distinct_count\": 1000\\n},\\n{\\n\"columns\": [\"nullable\"],\\n\"created_at\": \"2018-01-01 1:30:00.00000+00:00\",\\n\"row_count\": 1000,\\n\"distinct_count\": 10,\\n\"histo_buckets\": [\\n{\"num_eq\": 0, \"num_range\": 0, \"distinct_range\": 0, \"upper_bound\": \"-10\"},\\n{\"num_eq\": 0, \"num_range\": 1000, \"distinct_range\": 10, \"upper_bound\": \"0\"}\\n],\\n\"histo_col_type\": \"INT\"\\n},\\n{\\n\"columns\": [\"value\"],\\n\"created_at\": \"2018-01-01 1:30:00.00000+00:00\",\\n\"row_count\": 1000,\\n\"distinct_count\": 1000\\n}\\n]'"
    comparator.go:98: Executing "ALTER TABLE metrics INJECT STATISTICS e'[\\n{\\n\"columns\": [\"id\"],\\n\"created_at\": \"2018-01-01 1:00:00.00000+00:00\",\\n\"row_count\": 10,\\n\"distinct_count\": 10\\n},\\n{\\n\"columns\": [\"nullable\"],\\n\"created_at\": \"2018-01-01 1:30:00.00000+00:00\",\\n\"row_count\": 10,\\n\"distinct_count\": 10\\n},\\n{\\n\"columns\": [\"name\"],\\n\"created_at\": \"2018-01-01 1:30:00.00000+00:00\",\\n\"row_count\": 10,\\n\"distinct_count\": 10\\n}\\n]'"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON metric_id = id WHERE (\"time\" > '2020-01-01 00:00:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc INNER JOIN metrics ON metric_id = id WHERE (\"time\" > '2020-01-01 00:00:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values INNER JOIN metrics ON metric_id = id WHERE (\"time\" >= '2020-01-01 00:00:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc INNER JOIN metrics ON metric_id = id WHERE (\"time\" >= '2020-01-01 00:00:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values INNER JOIN metrics ON metric_id = id WHERE (\"time\" < '2020-01-01 00:00:00+00:00') AND (name = 'cpu')"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc INNER JOIN metrics ON metric_id = id WHERE (\"time\" < '2020-01-01 00:00:00+00:00') AND (name = 'cpu')"
    comparator.go:98: Executing "SELECT * FROM metric_values INNER JOIN metrics ON metric_id = id WHERE (\"time\" <= '2020-01-01 00:00:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc INNER JOIN metrics ON metric_id = id WHERE (\"time\" <= '2020-01-01 00:00:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values INNER JOIN metrics ON metric_id = id WHERE (\"time\" < '2020-01-01 00:00:10+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc INNER JOIN metrics ON metric_id = id WHERE (\"time\" < '2020-01-01 00:00:10+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values INNER JOIN metrics ON metric_id = id WHERE (\"time\" BETWEEN '2020-01-01 00:00:00+00:00' AND '2020-01-01 00:10:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc INNER JOIN metrics ON metric_id = id WHERE (\"time\" BETWEEN '2020-01-01 00:00:00+00:00' AND '2020-01-01 00:10:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metrics LEFT JOIN metric_values ON ((metric_id = id) AND (\"time\" BETWEEN '2020-01-01 00:00:00+00:00' AND '2020-01-01 00:10:00+00:00')) AND (name = 'cpu') ORDER BY value, id"
    comparator.go:98: Executing "SELECT * FROM metrics AS m WHERE EXISTS (SELECT * FROM metric_values AS mv WHERE (mv.metric_id = m.id) AND (\"time\" BETWEEN '2020-01-01 00:00:00+00:00' AND '2020-01-01 00:10:00+00:00')) ORDER BY m.id"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON (metric_id = id) AND (v.nullable = m.nullable) WHERE (\"time\" > '2020-01-01 00:00:00+00:00') AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable BETWEEN -20 AND -10) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable BETWEEN -20 AND -10) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable > 1) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable > 1) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable >= 1) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable >= 1) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable < -10) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable < -10) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable <= -10) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values_desc AS v INNER JOIN metrics AS m ON metric_id = id WHERE (v.nullable <= -10) AND (name = 'cpu') ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON metric_id = id WHERE ((\"time\" < '2020-01-01 00:00:10+00:00') AND (name = 'cpu')) AND (v.nullable = m.nullable) ORDER BY value"
    comparator.go:98: Executing "SELECT * FROM metric_values AS v INNER JOIN metrics AS m ON metric_id = id WHERE ((\"time\" < '2020-01-01 00:00:10+00:00') AND (name = 'cpu')) AND (v.nullable = 1) ORDER BY value"
    comparator.go:98: Executing "CREATE TABLE order_line (ol_o_id INT8, ol_i_id INT8); INSERT INTO order_line(ol_o_id, ol_i_id) VALUES (19, 6463), (20, 6463), (100, 6463), (101, 6463); CREATE INDEX ol_io ON order_line (ol_i_id, ol_o_id); CREATE TABLE stock (s_i_id INT8); INSERT INTO stock(s_i_id) VALUES (6463)"
    comparator.go:98: Executing "SELECT s_i_id, ol_o_id FROM stock INNER LOOKUP JOIN order_line ON s_i_id = ol_i_id WHERE ol_o_id BETWEEN 20 AND 100"
    comparator.go:98: Executing "SELECT s_i_id, ol_o_id FROM stock INNER LOOKUP JOIN order_line ON s_i_id = ol_i_id WHERE (ol_o_id IN (19, 20, 21)) AND (ol_o_id >= 20)"
    schemachanger_test.go:922: schema changer comparator testing succeeded on logic test "lookup_join_spans"
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestComparatorFromLogicTests.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-32164

Epic CRDB-30346

### PR DESCRIPTION
**Pull request by rafiss**
The tests previously worked by creating a logic test corpus then running
all those statements. However, that approach made it hard to parallelize
the tests using bazel test sharding.

Now, a test function is generated for each logic test file. Bazel will
take care of sharding all these tests.

fixes https://github.com/cockroachdb/cockroach/issues/111978
Release note: None

### CODE REVIEW

**Review by rickystewart — APPROVED** [review]

**Review by Xiang-Gu — COMMENTED** [review]
I didn't understand the purpose of first commit where we collect more than one corpus files -- it seems like the third commit just deleted all the corpus collection logic. Do we still need commit 1?

Also, in commit 2 (which seems like the meat of this PR), you use template to generate one test for each file, how is this going to be helped in terms of speeding up/parallelization? 

Finally, the comparator testing is not backported to 23.1 (it only exists on master and 23.2 branch), so I don't think we should backport this to 23.1?

**Comment by rafiss** [comment]
> I didn't understand the purpose of first commit where we collect more than one corpus files -- it seems like the third commit just deleted all the corpus collection logic. Do we still need commit 1?

You're right, we don't need it. That was my first attempt at speeding this up before I realized how we could use bazel's sharding.

> Also, in commit 2 (which seems like the meat of this PR), you use template to generate one test for each file, how is this going to be helped in terms of speeding up/parallelization?

Bazel will distribute each Test function in any package to a different shard. See https://cockroachlabs.slack.com/archives/C019A8NANH0/p1698258494931269 

> Finally, the comparator testing is not backported to 23.1 (it only exists on master and 23.2 branch), so I don't think we should backport this to 23.1?

I see, removing that label.

File: pkg/sql/schemachanger/comparator_generated_test.go [hunk 1]
```diff
@@ -0,0 +1,14316 @@
...
+
+	var path string
+	if bazel.BuiltWithBazel() {
+		var err error
+		path, err = bazel.Runfile(logicTestFile)
+		require.NoError(t, err)
+	} else {
+		path = "../../../" + logicTestFile
+	}
+
```
Comment (rickystewart) [inline pkg/sql/schemachanger/comparator_generated_test.go:14069]: So, first of all: you do not have to check this file in. Feel free to let it be `gitignore`'d. If you are going to check this in, I would encourage you to refactor out the massive duplication just so this file doesn't have have to be as colossal as it is.

**Comment by Xiang-Gu** [comment]
Thanks Rafi for your replies. It looks good to me now pending Ricky's comment.

File: pkg/sql/schemachanger/comparator_generated_test.go (hunk 1 above)
Comment (rafiss) [reply pkg/sql/schemachanger/comparator_generated_test.go]: i have removed the massive duplication

Comment (rickystewart) [reply pkg/sql/schemachanger/comparator_generated_test.go]: Oh, awesome, thank you. Now it just slows down my browser instead of crashing it. 😛

**Review by rickystewart — APPROVED** [review]

**Comment by rafiss** [comment]
tftr!

bors r+

**Pull request merged**